### ISSUE TITLE
Allow twist averaging with a single input file

### ISSUE THREAD

**Issue by jtkrogel** [QMCPACK/qmcpack#147]
Currently QMCPACK supports runs with multiple twists only with its multiple input file bundling facility.  The user currently has to create multiple files with differing twist number or twist angle that are otherwise identical.  This process is out of step with other input requests and places an unnecessary burden on the user.  This issue can be fixed by implementing new <simulationcell> inputs, such as <twist_numbers>/<twist_angles> elements (or equivalent hopefully standardized array inputs) that can be provided in a single input file.  Under the hood, QMCPACK would then split into the appropriate MPI groups as it currently does with a batched input file.

**Comment by prckent** [comment]
Indeed this would be convenient. Note that to implement this the MPI group initialization would have to be sufficiently delayed until after the XML inputs were sufficiently parsed.

**Comment by jtkrogel** [comment]
Yes, one would have to do a "look ahead" parse into <simulationcell> initially.

**Comment by jtkrogel** [comment]
If we are going to have a twist-averaged scalar file (consider stat.h5 later, I suppose), I would support writing it in addition to the per twist files, at least by default.

**Comment by rcclay** [comment]
Good idea.

**Comment by rcclay** [comment]
@ye-luo .  I don't know how best to handle the MPI...  One way to do it that would require minimal work to the QMCDrivers/BranchEngines is to stick to the way QMCPACK handles multiple input files...  the VMC/DMC runs should be totally independent.  Naively, one would think that this would take care of the orbital construction as well, but I don't know what would happen if you had a bunch of independent mpi processes trying to read the same hdf5 file.  Ideally, it would be nice to do an MPI_reduce or something on all the different EstimatorManagers to collect the twist averages, but I haven't thought precisely about how to structure this.

**Comment by jtkrogel** [comment]
I think the easiest thing is to just do a look ahead parse of the single input file, then if multiple twists are found, populate the stack of XML documents with copies (+change of twistnum) as is currently done when passing in a batched input file.  All the orbital generation, etc., would then just proceed in the usual way.

+1 on MPI_reduce over estimator managers (might have to think carefully about cloning/omp threads -- I believe Estimator manager follows the broader pattern of a master manager object + cloned slave object per thread).  It might also be best to act on per step data rather than per block for partial autocorrelation estimates (don't know what you have planned).

**Comment by ye-luo** [comment]
@rcclay no worry about h5. It is the same as now.
MPI_reduce is not difficult. The rank 0 of each twist processor pool does the work. A group leader communicator which I added on ANL branch can help.

**Comment by rcclay** [comment]
@ye-luo Great.  I'll take a look at your group communicator.  Setting up the higher level 

@jtkrogel With the current estimator manager, I was thinking of doing things per block.

**Comment by rcclay** [comment]
Great.  @ye-luo, Thanks.  I'll check out your new group leader communicator.

**Comment by jtkrogel** [comment]
Old
